src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,225 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace raksha::ir::auth_logic {
+
+class Principal {
+ public:
+  explicit Principal(std::string name) : name_(std::move(name)) {}
+  const std::string& name() const { return name_; }
+
+ private:
+  std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum Sign { kNegated, kPositive };
+
+// Predicate corresponds to a predicate in of the form 
+// <pred_name>(arg_1, ..., arg_n), and it may or may not be negated
+class Predicate {
+ public:
+  Predicate(std::string name, std::vector<std::string> args, Sign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        // TODO move surpresses an unused private field error about sign_
+        // for now, get rid of this eventually
+        sign_(std::move(sign)) {}
+
+  const std::string& name() const { return name_; }
+  const std::vector<std::string>& args() const { return args_; }
+  Sign sign() { return sign_; }
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// Attribute corresponds to an attribute of a principal with the form
+// <Principal> <pred_name>(arg_1, ..., arg_n)
+class Attribute {
+ public:
+  explicit Attribute(Principal principal, Predicate predicate)
+    : principal_(principal), predicate_(predicate) {}
+  const Principal& principal() const { return principal_; }
+  const Predicate& predicate() const { return predicate_; }
+ private:
+  Principal principal_;
+  Predicate predicate_;
+};
+
+// CanActAs corresponds to a canActAs expression of the form
+// <principal_1> canActAs <principal_2>
+class CanActAs {
+ public:
+  explicit CanActAs(Principal left_principal, Principal right_principal)
+    : left_principal_(left_principal),
+      right_principal_(right_principal) {}
```

Comment:
These initializations should use `std::move`.

Reply:
done